config.md
# Configuration

The main configuration, used for both writing and serving datasets, is specified in a toml file which
is exemplified and documented in the [sample config](config.sample.toml). Its path should be passed
in as the `NOZZLE_CONFIG` environment variable.

Configuring datasets to be extracted and served requires three different object storage directories:

- `dataset_defs_dir`: Contains the dataset definitions. This is the input to the extraction process.
- `providers_dir`: Auxiliary to the dataset definitions, configures providers for external services
  like Firehose and Substreams.
- `data_dir`: Where the actual dataset parquet tables are stored once extracted. Can be initially empty.

Although the initial setup with three directories may seem cumbersome, it allows for a highly
flexible configuration.

# Using env vars

Note that the values in the `NOZZLE_CONFIG` file can be overridden from the environment, by prefixing
the env var name with `NOZZLE_CONFIG_`. For example, to override the `data_dir` value, you can set a
`NOZZLE_CONFIG_DATA_DIR` env var to the desired path.

# Service addresses

The following optional configuration keys allow you to control the hostname and port that each service binds to:

- `flight_addr`: Arrow Flight RPC server address (default: `0.0.0.0:1602`)
- `jsonl_addr`: JSON Lines server address (default: `0.0.0.0:1603`)  
- `registry_service_addr`: Registry service address (default: `0.0.0.0:1611`)
- `admin_api_addr`: Admin API server address (default: `0.0.0.0:1610`)

## Logging

Simplified control of the logging verbosity level is offered by the `NOZZLE_LOG` env var. It accepts
the values `error`, `warn`, `info`, `debug` or `trace`. The default value is `debug`. The standard
`RUST_LOG` env var can be used for finer-grained log filtering.

# Configuring object stores

All directory configurations (the `*_dir` keys) support both filesystem and object store locations.
So they can either be a filesystem path, for local storage, or a URL for object an store. For
production usage, an object store is recommended. Object store URLs can be in one of the following
formats:

### S3-compatible stores

URL format: `s3://<bucket>`

Sessions can be configured through the following environment variables:

- `AWS_ACCESS_KEY_ID`: access key ID
- `AWS_SECRET_ACCESS_KEY`: secret access key
- `AWS_DEFAULT_REGION`: AWS region
- `AWS_ENDPOINT`: endpoint
- `AWS_SESSION_TOKEN`: session token
- `AWS_ALLOW_HTTP`: allow non-TLS connections

### Google Cloud Storage (GCS)

URL format: `gs://<bucket>`

GCS Authorization can be configured through one of the following environment variables:

- `GOOGLE_SERVICE_ACCOUNT_PATH`: location of service account file, or
- `GOOGLE_SERVICE_ACCOUNT_KEY`: JSON serialized service account key, or
- Application Default Credentials.

## Datasets

All datasets have a name. This will be used as the dataset directory name under the data directory,
and also as the catalog schema name in the SQL interface. So if you have a dataset name `foobar`, it
will by default be placed under `<data_dir>/foobar/` and tables will be refered to in SQL as
`foobar.table`.

Conceptually there are raw datasets, which are extracted from external systems such Firehose and
Substreams, and then there are datasets defined as views on other datasets.

## Raw datasets

Details for the raw datasets currently implemented:

- EVM RPC [dataset docs](evm-rpc-datasets/README.md)
- Firehose [dataset docs](firehose-datasets/README.md)
- Substreams [dataset docs](substreams-datasets/README.md)

## Datasets

A dataset can be defined as a set of queries on other datasets, with each query defining a view. Dataset views can be materialized, to share a result
among multiple queries or to do ahead of time work that is too slow or expensive to do at query time.
As an example, we will show how to define an `erc20_transfer` dataset, containing all erc20 transfer
events. The `erc20_transfer.toml` is straightforward:

```
kind = "sql"
name = "erc20_transfer"
```

The actual SQL queries should be placed beside the dataset definition, in a directory of the same
name. In this case we will only have one query for the transfers table, so the file `erc_20_transfer/transfers.sql` would contain the query:

```sql
SELECT 
    t.block_num,
    t.timestamp,
    t.event['from'] AS from,
    t.event['to'] AS to,
    t.event['value'] AS value
FROM (
    SELECT 
        l.block_num,
        l.timestamp,
        evm_decode_log(l.topic1, l.topic2, l.topic3, l.data, 'Transfer(address indexed from, address indexed to, uint256 value)') AS event
    FROM eth_firehose.logs l
    WHERE
        l.topic0 = evm_topic('Transfer(address indexed from, address indexed to, uint256 value)')
        and l.topic3 is null
) t;

```

The dataset is now ready to be dumped with `nozzle dump --dataset erc_20_transfer`, assuming the dependency
`eth_firehose` is already present, and then queried as in `select * from erc_20_transfer.transfers`.


common/src/config.rs
use std::{net::SocketAddr, path::PathBuf, sync::Arc};

use datafusion::{
    error::DataFusionError,
    execution::{
        disk_manager::{DiskManagerBuilder, DiskManagerMode},
        memory_pool::{FairSpillPool, GreedyMemoryPool, MemoryPool},
        runtime_env::RuntimeEnvBuilder,
    },
};
use figment::{
    Figment,
    providers::{Env, Format as _, Toml},
};
use fs_err as fs;
use js_runtime::isolate_pool::IsolatePool;
use metadata_db::{KEEP_TEMP_DIRS, MetadataDb, temp_metadata_db};
use serde::Deserialize;
use thiserror::Error;

use crate::{Store, query_context::QueryEnv};

#[derive(Debug, Clone)]
pub struct Config {
    pub data_store: Arc<Store>,
    pub providers_store: Arc<Store>,
    pub dataset_defs_store: Arc<Store>,
    pub metadata_db_url: String,
    pub max_mem_mb: usize,
    pub spill_location: Vec<PathBuf>,
    pub microbatch_max_interval: u64,
    /// Addresses to bind the server to. Used during testing.
    pub addrs: Addrs,
    pub config_path: PathBuf,
}

#[derive(Debug, Clone)]
pub struct Addrs {
    pub flight_addr: SocketAddr,
    pub jsonl_addr: SocketAddr,
    pub registry_service_addr: SocketAddr,
    pub admin_api_addr: SocketAddr,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ConfigFile {
    pub data_dir: String,
    pub providers_dir: String,
    pub dataset_defs_dir: String,
    pub metadata_db_url: Option<String>,
    pub max_mem_mb: usize,
    pub spill_location: Vec<PathBuf>,
    pub microbatch_max_interval: Option<u64>,
    pub flight_addr: Option<String>,
    pub jsonl_addr: Option<String>,
    pub registry_service_addr: Option<String>,
    pub admin_api_addr: Option<String>,
}

pub type FigmentJson = figment::providers::Data<figment::providers::Json>;

#[derive(Error, Debug)]
pub enum ConfigError {
    #[error("IO error at {0}: {1}")]
    Io(PathBuf, std::io::Error),
    #[error("Missing required config at {0}: {1}")]
    MissingConfig(PathBuf, &'static str),
    #[error("Config parse error at {0}: {1}")]
    Figment(PathBuf, figment::Error),
    #[error("Store error at {0}: {1}")]
    Store(PathBuf, crate::store::StoreError),
    #[error("Metadata DB error at {0}: {1}")]
    MetadataDb(PathBuf, metadata_db::Error),
    #[error("Invalid address format for {0}: {1}")]
    InvalidAddress(String, String),
}

impl Config {
    ///
    /// `env_override` allows env vars prefixed with `NOZZLE_CONFIG_` to override config values.
    pub async fn load(
        file: impl Into<PathBuf>,
        env_override: bool,
        config_override: Option<Figment>,
        allow_temp_db: bool,
    ) -> Result<Self, ConfigError> {
        let input_path = file.into();
        let config_path = fs::canonicalize(&input_path)
            .map_err(|err| ConfigError::Io(input_path.clone(), err))?;
        let contents = fs::read_to_string(&config_path)
            .map_err(|err| ConfigError::Io(config_path.clone(), err))?;

        let config_file: ConfigFile = {
            let mut config_builder = Figment::new().merge(Toml::string(&contents));
            if env_override {
                config_builder = config_builder.merge(Env::prefixed("NOZZLE_CONFIG_"));
            }
            if let Some(config_override) = config_override {
                config_builder = config_builder.merge(config_override);
            }
            config_builder
                .extract()
                .map_err(|e| ConfigError::Figment(config_path.clone(), e))?
        };

        // Resolve any filesystem paths relative to the directory of the config file.
        let base = config_path.parent();
        let addrs = Addrs::from_config_file(&config_file, Addrs::default())?;
        let data_store = Store::new(config_file.data_dir, base)
            .map_err(|e| ConfigError::Store(config_path.clone(), e))?;
        let providers_store = Store::new(config_file.providers_dir, base)
            .map_err(|e| ConfigError::Store(config_path.clone(), e))?;
        let dataset_defs_store = Store::new(config_file.dataset_defs_dir, base)
            .map_err(|e| ConfigError::Store(config_path.clone(), e))?;

        let metadata_db_url = match (config_file.metadata_db_url, allow_temp_db) {
            (Some(url), _) => url,
            (None, true) => temp_metadata_db(*KEEP_TEMP_DIRS).await.url().to_string(),
            (None, false) => {
                return Err(ConfigError::MissingConfig(
                    config_path.clone(),
                    "metadata_db_url",
                ));
            }
        };

        Ok(Self {
            data_store: Arc::new(data_store),
            providers_store: Arc::new(providers_store),
            dataset_defs_store: Arc::new(dataset_defs_store),
            metadata_db_url,
            max_mem_mb: config_file.max_mem_mb,
            spill_location: config_file.spill_location,
            microbatch_max_interval: config_file.microbatch_max_interval.unwrap_or(100_000),
            addrs,
            config_path,
        })
    }

    pub fn make_query_env(&self) -> Result<QueryEnv, DataFusionError> {
        use datafusion::execution::cache::{
            cache_manager::CacheManagerConfig, cache_unit::DefaultFileStatisticsCache,
        };

        let spill_allowed = !self.spill_location.is_empty();
        let disk_manager_mode = if spill_allowed {
            DiskManagerMode::Disabled
        } else {
            DiskManagerMode::Directories(self.spill_location.clone())
        };

        let disk_manager_builder = DiskManagerBuilder::default().with_mode(disk_manager_mode);

        let memory_pool: Option<Arc<dyn MemoryPool>> = if self.max_mem_mb > 0 {
            let max_mem_bytes = self.max_mem_mb * 1024 * 1024;

            if spill_allowed {
                Some(Arc::new(FairSpillPool::new(max_mem_bytes)))
            } else {
                Some(Arc::new(GreedyMemoryPool::new(max_mem_bytes)))
            }
        } else {
            None
        };
        let cache_manager = CacheManagerConfig {
            table_files_statistics_cache: Some(Arc::new(DefaultFileStatisticsCache::default())),
            list_files_cache: None,
        };

        let mut runtime_config = RuntimeEnvBuilder::new()
            .with_disk_manager_builder(disk_manager_builder)
            .with_cache_manager(cache_manager);

        if let Some(memory_pool) = memory_pool {
            runtime_config = runtime_config.with_memory_pool(memory_pool);
        }

        let runtime_env = runtime_config.build()?;
        let isolate_pool = IsolatePool::new();
        Ok(QueryEnv {
            df_env: Arc::new(runtime_env),
            isolate_pool,
        })
    }

    pub async fn metadata_db(&self) -> Result<MetadataDb, ConfigError> {
        MetadataDb::connect(&self.metadata_db_url)
            .await
            .map_err(|e| ConfigError::MetadataDb(self.config_path.clone(), e))
    }
}

impl Default for Addrs {
    fn default() -> Self {
        Self {
            flight_addr: ([0, 0, 0, 0], 1602).into(),
            jsonl_addr: ([0, 0, 0, 0], 1603).into(),
            registry_service_addr: ([0, 0, 0, 0], 1611).into(),
            admin_api_addr: ([0, 0, 0, 0], 1610).into(),
        }
    }
}

impl Addrs {
    pub fn from_config_file(
        config_file: &ConfigFile,
        default_addrs: Addrs,
    ) -> Result<Self, ConfigError> {
        let parse_addr = |addr_str: &Option<String>,
                          default: SocketAddr,
                          name: &str|
         -> Result<SocketAddr, ConfigError> {
            match addr_str {
                Some(addr) => addr
                    .parse::<SocketAddr>()
                    .map_err(|e| ConfigError::InvalidAddress(name.to_string(), e.to_string())),
                None => Ok(default),
            }
        };

        Ok(Self {
            flight_addr: parse_addr(
                &config_file.flight_addr,
                default_addrs.flight_addr,
                "flight_addr",
            )?,
            jsonl_addr: parse_addr(
                &config_file.jsonl_addr,
                default_addrs.jsonl_addr,
                "jsonl_addr",
            )?,
            registry_service_addr: parse_addr(
                &config_file.registry_service_addr,
                default_addrs.registry_service_addr,
                "registry_service_addr",
            )?,
            admin_api_addr: parse_addr(
                &config_file.admin_api_addr,
                default_addrs.admin_api_addr,
                "admin_api_addr",
            )?,
        })
    }
}
